### PR TITLE
Avoiding expensive hash computation in filter ranking strategy

### PR DESCRIPTION
**Pull request by steigma**
Avoiding expensive hash computation for complex steps with many children in filter ranking strategy by using identity hash map instead of linked hash map. With the linked hash map, a hash computation is repeated for the traversal parent with each child when doing the grouping, which can be quite costly when there are several hundreds of children (especially since the hash computation also includes the hash of child steps).

### CODE REVIEW

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/FilterRankingStrategy.java [hunk 1]
```diff
@@ -232,4 +243,64 @@ public Set<Class<? extends OptimizationStrategy>> applyPrior() {
...
+
+    /**
+     * Get steps of the specified class throughout the traversal and grouping them based on the traversal parent
+     * collecting them in a fashion that orders them from the deepest steps first
+     */
+    public static List<Pair<TraversalParent, List<Step<?,?>>>> collectStepsOfAssignableClassRecursivelyFromDepthGroupedByParent(
+            final Traversal.Admin<?, ?> traversal, final Class<?> stepClass) {
+
+        final List<Pair<TraversalParent, List<Step<?,?>>>> collectionList = new ArrayList<>();
+        final Stack<Step<?,?>> stack = new Stack<>();
```
Comment (vkagamlyk) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/FilterRankingStrategy.java:255]: nit: not informative variable names. may be something like `unprocessedSteps` will be better?

**Comment by vkagamlyk** [comment]
Good to have some test cases, especially for traversal where there are steps with nested global and local children

**Comment by Cole-Greer** [comment]
I agree with @vkagamlyk that some additional tests would be beneficial here. It is quite hard to prove that `collectStepsOfAssignableClassRecursivelyFromDepthGroupedByParent()` does not effect the order of the resulting map in a meaningful way through reasoning alone. I would also be curious to know how substantial the performance difference can be in a "worst case" scenario with hundreds of child steps.

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/FilterRankingStrategyTest.java [hunk 2]
```diff
@@ -85,44 +86,148 @@ public void doTest() {
...
-                {has("value", 0).filter(out()).dedup(), has("value", 0).filter(out()).dedup(), Collections.emptyList()},
-                {has("value", 0).or(has("name"), has("age")).has("value", 1).dedup(), has("value", 0).has("value", 1).or(has("name"), has("age")).dedup(), Collections.singletonList(InlineFilterStrategy.instance())},
-                {has("value", 0).or(out(), in()).as(Graph.Hidden.hide("x")).has("value", 1).dedup(), has("value", 0).has("value", 1).or(outE(), inE()).dedup(), TraversalStrategies.GlobalCache.getStrategies(Graph.class).toList()},
-                {has("value", 0).and(has("age"), has("name", "marko")).is(10), __.is(10).has("value", 0).has("age").has("name", "marko"), Collections.singletonList(InlineFilterStrategy.instance())},
-                {has("value", 0).filter(or(not(has("age")), has("age", 1))).has("value", 1).dedup(), has("value", 0).has("value", 1).or(not(filter(properties("age"))), has("age", 1)).dedup(), TraversalStrategies.GlobalCache.getStrategies(Graph.class).toList()},
```
Comment (steigma) [inline gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/FilterRankingStrategyTest.java:125]: I reformatted these tests slightly and added some new ones on the top

**Comment by steigma** [comment]
Thanks for the feedback. I added a few tests and renamed some variables. I also found that the lambda handling was not really working since the step collection was restricted to TraversalParent subclasses. This should now also be fixed and I added some basic test coverage for this. 

In our experiments with queries of the form
```
g.V('vertex')...
coalesce(
    and(
      values('property').is('val1'),
      values('property').is('val2'),
      ...
      values('property').is('val800')
   ),
   sideEffect(properties('property').drop()).
   property(set, 'propery', 'val1').
   property(set, 'propery', 'val2').
   ...
   property(set, 'propery', 'val800')
)...
```
we have seen improvement of around 25% including all strategies and executing the queries, i.e,. the FilterRankingStrategy improved significantly (as also other strategies taking some significant amount of time).

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/FilterRankingStrategyCacheInitializationTest.java [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+ * Testing whether the cache for the FilterRankingStrategy is correctly initialized,
+ * which includes gathering/collecting the traversal parents and their children steps
+ * in the correct deep first manner.
+ */
+@RunWith(JUnit4.class)
+public class FilterRankingStrategyCacheInitializationTest {
+
+
+    @Test
+    public void test_collection_lambda_nested() {
```
Comment (vkagamlyk) [inline gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/FilterRankingStrategyCacheInitializationTest.java:58]: nit: test name

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/FilterRankingStrategyCacheInitializationTest.java [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+                __.where(
+                        __.or(
+                                __.select("n").hasLabel("software"),
+                                __.select("n").map(Lambda.function("it.get().out()"))
+                        )
+                ));
+
+        // cache entries
+        final List<Pair<Boolean, Set<String>>> expectedTraversalParentCacheEntries = Arrays.asList(
+                new Pair<>(true, new HashSet<>(Collections.emptyList())),
```
Comment (vkagamlyk) [inline gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/FilterRankingStrategyCacheInitializationTest.java:84]: Why need to pass `Collections.emptyList()` to `HashSet` constructor?

**Comment by kenhuuu** [comment]
I don't completely follow the changes being made here but they seem to be well tested so my VOTE is +1.

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/FilterRankingStrategyCacheInitializationTest.java (hunk 3 above)
Comment (steigma) [reply gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/FilterRankingStrategyCacheInitializationTest.java]: Not completely sure what you mean, but I tried to improve the names of the test cases.

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/FilterRankingStrategyCacheInitializationTest.java (hunk 4 above)
Comment (steigma) [reply gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/FilterRankingStrategyCacheInitializationTest.java]: It is not necessary. I removed it in the new revision. Also changed some arguments to Optional to make it clear that they do not need to be provided (where we then just compare with the old cache initialization method).

File: gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/FilterRankingStrategyCacheInitializationTest.java (hunk 3 above)
Comment (vkagamlyk) [reply gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/FilterRankingStrategyCacheInitializationTest.java]: test naming convention is camelCase and `shouldDescribeWhatTestDo`.
please look at the names of other tests

Comment (steigma) [reply gremlin-core/src/test/java/org/apache/tinkerpop/gremlin/process/traversal/strategy/optimization/FilterRankingStrategyCacheInitializationTest.java]: Thanks, I fixed the test names in the new version

**Comment by vkagamlyk** [comment]
Thank you @steigma!
VOTE +1

**Comment by Cole-Greer** [comment]
Thanks @steigma, the new tests LGTM. VOTE +1

**Pull request merged**